embedded-graphics/src/primitives/rectangle.rs

```diff
@@ -139,6 +139,104 @@ impl Rectangle {
+    ///
+    /// If no intersection is present, this method will return a zero sized rectangle with its top
+    /// left corner set to `(0, 0)`.
+    ///
+    /// # Examples
+    ///
+    /// ## Intersection
+    ///
+    /// This example draws two rectangles to a mock display using the `.` character, along with
+    /// their intersection shown with `#` characters.
+    ///
+    /// ```rust
+    /// use embedded_graphics::{
+    ///     mock_display::MockDisplay, pixelcolor::BinaryColor, prelude::*, primitives::Rectangle,
+    ///     style::PrimitiveStyle,
+    /// };
+    ///
+    /// let mut display = MockDisplay::new();
+    /// # display.set_allow_overdraw(true);
+    ///
+    /// let rect1 = Rectangle::new(Point::zero(), Size::new(7, 8));
+    /// let rect2 = Rectangle::new(Point::new(2, 3), Size::new(10, 7));
+    ///
+    /// let intersection = rect1.intersection(&rect2);
+    ///
+    /// rect1
+    ///     .into_styled(PrimitiveStyle::with_stroke(BinaryColor::Off, 1))
+    ///     .draw(&mut display)?;
+    ///
+    /// rect2
+    ///     .into_styled(PrimitiveStyle::with_stroke(BinaryColor::Off, 1))
+    ///     .draw(&mut display)?;
+    ///
+    /// intersection
+    ///     .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 1))
+    ///     .draw(&mut display)?;
+    ///
+    /// assert_eq!(
+    ///     display,
+    ///     MockDisplay::from_pattern(&[
+    ///         ".......     ",
+    ///         ".     .     ",
+    ///         ".     .     ",
+    ///         ". #####.....",
+    ///         ". #   #    .",
+    ///         ". #   #    .",
+    ///         ". #   #    .",
+    ///         "..#####    .",
+    ///         "  .        .",
+    ///         "  ..........",
+    ///     ])
+    /// );
+    /// # Ok::<(), core::convert::Infallible>(())
+    /// ```
+    ///
+    /// ## No intersection
+    ///
+    /// This example creates two rectangles with no intersection between them. In this case,
+    /// `intersection` returns a zero-sized rectangle with its origin at `(0, 0)`.
+    ///
+    /// ```rust
+    /// use embedded_graphics::{
+    ///     mock_display::MockDisplay, pixelcolor::BinaryColor, prelude::*, primitives::Rectangle,
+    ///     style::PrimitiveStyle,
+    /// };
+    ///
+    /// let mut display: MockDisplay<BinaryColor> = MockDisplay::new();
+    ///
+    /// let rect1 = Rectangle::new(Point::zero(), Size::new(7, 8));
+    /// let rect2 = Rectangle::new(Point::new(10, 15), Size::new(10, 7));
+    ///
+    /// let intersection = rect1.intersection(&rect2);
+    ///
+    /// assert_eq!(intersection.size, Size::zero());
+    /// assert_eq!(intersection.top_left, Point::zero());
+    /// # Ok::<(), core::convert::Infallible>(())
+    /// ```
+    pub fn intersection(&self, other: &Rectangle) -> Rectangle {
+        let other_bottom_right = other.bottom_right().unwrap_or_else(Point::zero);
+        let self_bottom_right = self.bottom_right().unwrap_or_else(Point::zero);
```

Comment:
I think this contains an error and this test should show the problem (haven't tried to run it):
```rust
let rect1 = Rectangle::new(Point::new(0, 0), Size::new(0, 0));
let rect2 = Rectangle::new(Point::new(-10, -10), Size::new(20, 20));

assert_eq!(rect1.intersection(&rect2).size, Size::zero());
```
If any `bottom_right` is `None` the intersection must always be `Size::zero()`.

Reply:
That is indeed broken, well spotted. Should we add the size check to `Rectangle`'s `ContainsPoint::contains` impl as well?

Reply:
I don't think that `contains` needs an additional size check because it checks the  result of `bottom_right`, that is `None` if the rectangle is zero sized. But we could use the same check for `intersection`:
```rust
    pub fn intersection(&self, other: &Rectangle) -> Rectangle {
        if let (Some(other_bottom_right), Some(self_bottom_right)) =
            (other.bottom_right(), self.bottom_right())
        {
            // Check for overlap
            if self.contains(other.top_left)
                || self.contains(other_bottom_right)
                || other.contains(self.top_left)
                || other.contains(self_bottom_right)
            {
                return Rectangle::with_corners(
                    self.top_left.component_max(other.top_left),
                    self_bottom_right.component_min(other_bottom_right),
                )
            }
        }

        // No overlap present
        Rectangle::new(Point::zero(), Size::zero())
    }
```